rust/rope/src/rope.rs

```diff
@@ -265,12 +269,82 @@ impl Metric<RopeInfo> for LinesMetric {
     fn can_fragment() -> bool { true }
 }
 
+#[derive(Clone, Copy)]
+pub struct Utf16CodeUnitsMetric(usize);
+
+impl Metric<RopeInfo> for Utf16CodeUnitsMetric {
+    fn measure(info: &RopeInfo, _: usize) -> usize {
+        info.utf16_size
+    }
+
+    fn is_boundary(s: &String, offset: usize) -> bool {
+        s.is_char_boundary(offset)
+    }
+
+    fn to_base_units(s: &String, in_measured_units: usize) -> usize {
+        let mut cur_len_utf16 = 0;
+        let mut cur_len_utf8 = 0;
+        for u in s.chars() {
+            cur_len_utf16 += u.len_utf16();
+            cur_len_utf8 += u.len_utf8();
+            if cur_len_utf16 >= in_measured_units {
```

Comment:
@cmyr I need this condition to get it to work properly. After using it works perfectly for cases I tested i.e. text with and without non-ASCII characters. However, I am not sure why should the `>=` be needed and not just `==`.

Reply:
Where are those test cases??

Reply:
In any case: `==` would fail if, (say) `s` was `😈` and `in_measured_units` was `1`, because `"😈".len_utf16() == 2`.

Reply:
Yes, this sounds right to me, at least it's consistent with the lines metric. When the measured units are not on a boundary, I think `to_base_units` should "round down". I'm not sure if this is documented anywhere, but as I say being consistent with lines is good.